### ISSUE TITLE
Multi-plant parallelization 

### ISSUE THREAD

**Issue by vtrifunovic** [danforthcenter/plantcv#661]
Hello, 
I've been trying to run PlantCV on a Raspberry Pi 3B+ and I've been having issues using the plantcv-workflow.py to parallel process the multi_plant_tutorial.py workflow. For a single plant workflow the parallelization works just fine, but when I first ran it with the multi plant workflow the final CSV output had no data other than the metadata. 

I assumed this was because plantcv-workflow.py didn't know what result file to pull data from since the tutorial code outputs separate text files for each plant. To try and fix this I added a parser argument to the top of the code for a "-r" result file, and set print results to
`pcv.print_results(filename = args.result)`
instead of the 
`pcv.print_results(filename = 'prefix_' + str(i) + '.txt.')`
that was originally there.

This did produce an output with the parallelization, and covering it to a CSV file I got:
![multiplantcsv](https://user-images.githubusercontent.com/76224191/102653373-7c08fa00-4134-11eb-8932-a07b36c1c0ce.png)
(same image was used twice, thats why there's two rows of data)

This only places the output of the last plant processed in the picture (makes sense since the: `for i in range(0, len(rois1))` loop would set the last plant as the args.result file), and testing this workflow independently my results.txt file did only have one line of data. To get the data from all the plants I edited the workflow to create another text file that would be defined as args.result to record the data from each individual plant file, and placed this inside the previously mentioned loop:
```pcv.print_results(filename = 'prefix_' + str(i) + '.txt')
with open(args.result, "a") as finalfile:
    with open('prefix_' + str(i) + '.txt') as tempfile:
        for x in tempfile.readlines():
            finalfile.write(x)
            finalfile.write("\n")
    tempfile.close()
    os.remove('prefix_' +str(i)+ '.txt')
finalfile.close()
pcv.outputs.clear()
```

Testing this workflow independently the result file did contain all of the data that would have been separated over the other text files, but it refuses to run with plantcv-workflow.py and shoots a ".JSONDecodeError: Extra data" error. I have tried messing with the process_results file to see if I can have it decode the data on a line-by-line basis, but no matter what I do it keeps giving me the same error. 

Now I am wondering if I have been running something wrongly this entire time or if there's a simple fix to my issue.

Final note: I know that the multi_plant_tutorial.py was intended to run with a --names file, but I haven't been able to figure out how to work that or if that's what's causing my parallelization issue.

**Comment by nfahlgren** [comment]
Hi @vtrifunovic!

You have definitely identified a part of multi-plant workflows that is still not polished as much as we would like. Looking back, the example workflow as given in our documentation is not set up for parallelization, but the solution you have is almost exactly what you need.

Just to give a little more insight into what is going on, `plantcv-workflow.py` scans the input directory for images and it creates a temporary results file (a JSON file with image metadata) for each image. When your multi-plant workflow runs, it does need the `args.result` input that you added to work with `plantcv-workflow.py`. For a single-plant workflow, you would simply run `pcv.print_results(args.result)` and the data would get recorded to the temporary JSON file. Then at the end `plantcv-workflow.py` would combine all the temporary JSON files into one final JSON file.

The way we currently build the temporary results files only allows for one set of results/observations. Since the multi-plant images have more than one set of results, we need to read the metadata from the temporary file created by `plantcv-workflow.py`, then delete that file. Then for each plant, make a new result file with that metadata duplicated and use the new one to save results for each plant. Almost exactly what you have, just on separate result files.

This Jupyter notebook uses multiple ROIs instead of the `pcv.cluster_contours` function, but in the last two cells I extract the image metadata and create new result files for each plant: https://github.com/nfahlgren/shared_jupyter_notebooks/blob/master/arabidopsis-multiplant.ipynb. If any of that is not clear, let me know, happy to help!

**Comment by vtrifunovic** [comment]
I think I did something wrong when I tried to apply what you did in the last two cells of your notebook. At first I wasn't able to get it to work as a parallel workflow, since I encountered a FileNotFoundError, which I solved by changing `filename = str(i) + "_" + args.result` to `filename = args.result + "_" + str(i)`. After that the code worked, but produced an empty json. 
When I changed filename to `args.result` it produced the following CSV table:
![multiplantcsv2](https://user-images.githubusercontent.com/76224191/102822559-e9b85e80-439e-11eb-92f9-50649e9676c1.png)
And running it as a single image workflow produces 18 correctly named files. 

Based on your explanation of `plantcv-workflow.py` I think I understand now how it works, but I'm not sure how to get it to loop over each plant in the picture to read and then delete the temp/result files.

**Comment by nfahlgren** [comment]
Hi @vtrifunovic, if possible, could you upload your Jupyter notebook or script here? Happy to take a look. Note that ipynb files cannot be uploaded to GitHub, but if you zip it first it will let you.

**Comment by vtrifunovic** [comment]
Here's the workflow I was using. I've been running PlantCV using the raspberry pi terminal 
[multi_plant_tutorial.zip](https://github.com/danforthcenter/plantcv/files/5730745/multi_plant_tutorial.zip)

**Comment by nfahlgren** [comment]
I think the issue is in the for loop at the end of the script. `object_composition` and `analyze_object` are run twice, the second time it's run it will overwrite the results of the first time. The first time it's run it's on the individual plant but the second time it's on all the plants at once. So I think if you commented out lines 319 and 321 it would work. But definitely let me know if that's not the fix!

**Comment by vtrifunovic** [comment]
I commented lines 319 and 321 out, but the program still produced an empty json file (when I say empty I mean all that is in the output json is: `{"variables": {}, "entities": []}` ). I also noticed that I still had a few sections where `filename` was supposed to be, auto-replaced with `filename2` from previous attempts at editing the workflow, but after correcting that the program still produced the same result.

**Comment by vtrifunovic** [comment]
I figured out what my issue was. After changing `filename = str(i) + "_" + args.result` to `filename = args.result + "_" + str(i)`, the ".txt" naturally produced by args.result at the end was lost. I'm not sure what type of file it became, but it wasn't something that the rest of the parallelization could recognize. Once I changed the line to say `filename = args.result + "_" + str(i) + ".txt"` the parallelization worked perfectly!
